rootfs/etc/nginx/lua/balancer.lua

```diff
@@ -122,28 +98,46 @@ function _M.init_worker()
   end
 end
 
-function _M.call()
-  local phase = ngx.get_phase()
-  if phase == "log" then
-    after_balance()
-    return
+function _M.rewrite()
+  local balancer = get_balancer()
+  if not balancer then
+    ngx.status = ngx.HTTP_SERVICE_UNAVAILABLE
+    return ngx.exit(ngx.status)
   end
-  if phase ~= "balancer" then
-    return error("must be called in balancer or log, but was called in: " .. phase)
+end
+
+function _M.balance()
+  local balancer = get_balancer()
+  if not balancer then
+    return
   end
 
-  local host, port = balance()
+  local host, port = balancer:balance()
   if not host then
-    ngx.status = ngx.HTTP_SERVICE_UNAVAILABLE
-    return ngx.exit(ngx.status)
+    ngx.log(ngx.WARN, "no host returned, balancer: " .. balancer.name)
```

Comment:
I'm curious why you got rid of the 503 here? To me it seems like it would be better to be more explicit about this error case.

Reply:
as mentioned at https://github.com/kubernetes/ingress-nginx/pull/2543/files#r191262645 it is not possible to modify status at this phase.